### ISSUE TITLE
SBT plugins seems to share ClassLoader and classes with SBT

### ISSUE THREAD

**Issue by kanischev** [sbt/sbt#4000]
Hello!
I'm trying to make sbt plugin to generate Slick codegen sources for DB Schema for latest PlayFramework(2.6.12) application and sbt v. 1.1.1.
The idea was to set up in-memory db, apply application's migrations on it and run slick-codegen.
There is tototoshi's plugin that works a little bit different: it uses existing db connection to get schema and I decided to hack SBT a little.
I've tried few variants and neither works as expected: they all say "java.lang.NoSuchMethodError: com.typesafe.config.Config.getIsNull(Ljava/lang/String;)"
## steps
1. Task in build.sbt

a) created PlayFramework 2.6.12 project
b) added module SlickCodegen
c) specified dependencies in plugins.sbt (including a new version of com.typesafe.config)
d) created build.sbt for module:

```
import java.nio.file.FileSystems
import Dependencies._
import com.typesafe.config.ConfigFactory
import play.api
import play.api.test.Helpers
import play.api.{Environment, Mode}
import play.api.db.Databases
import play.api.db.evolutions.{EnvironmentEvolutionsReader, Evolutions, ResourceEvolutionsReader}
import play.api.inject.guice.GuiceApplicationBuilder
import slick.codegen.SourceCodeGenerator
import scala.compat.Platform

libraryDependencies ++= slickModuleDependencies

lazy val generateSchema = taskKey[Seq[File]]("Generates slick representation classes for actual schema (runned on embedded database)")


generateSchema := {
  val outputDir = "src/main/scala"
  val pkg = "models"

  val app = new GuiceApplicationBuilder()
    .in(new Environment(baseDirectory.value.getParentFile.getParentFile / "conf" / "dev.conf",
      classOf[GuiceApplicationBuilder].getClassLoader,
      Mode.Test)
    ).build()
  Helpers.running(new GuiceApplicationBuilder().build()){
    Databases.withInMemory(urlOptions = Map(
      "MODE" -> "PostgreSQL"
    )) { database =>
      Evolutions.withEvolutions(database, new ResourceEvolutionsReader{
        def loadResource(db: String, revision: Int) = {
          Option(new File(
            baseDirectory.value.getParentFile.getParentFile, "conf/evolutions/default")).filter(_.exists()).map(f => java.nio.file.Files.newInputStream(f.toPath))
        }
      }) {
        println("Evolutions applied!")
        SourceCodeGenerator.main(
          Array(
            "slick.jdbc.PostgresProfile",
            "org.h2.Driver",
            database.url,
            outputDir,
            pkg,
            "",
            "")
        )
      }
    }
  }
  println("Tables source generated")
  val fname = outputDir + s"/${pkg.replace('.', '/')}/Tables.scala"
  Seq(file(fname))
}

sourceGenerators in Compile += generateSchema.taskValue
```

2. Created separate SBT AutoPlugin:
[plugin in my repo](https://github.com/kanischev/sbt-play-slick-codegen)
Tried to use it with no luck - the same problem occurres

## problem
It looks like com.typesafe.config of 1.2.0 is the dependency of SBT (actually, I've found it in sbt/lib/local-preloaded directory of sbt home) and is loaded before any of SBT plugins and build script. plugins.sbt libraryDependencies does not cause newer version to be used. plugins use the same classloader - though their dependencies can interfere causing different classloading issues.
Are there any workarounds?

## expectation
I expected plugins would be loaded by separate ClassLoaders to minimize problems with different libraries versions - but they are not.

## notes

sbt version: *1.1.* *

**Comment by dwijnand** [comment]
> I expected plugins would be loaded by separate ClassLoaders to minimize problems with different libraries versions - but they are not.

no, plugins are meant to allow building one on another (e.g sbt-play on sbt-web).

on that basis I'll close this ticket (which, however, doesn't prevent further comments on the topic! 😃)

**Comment by kanischev** [comment]
Maybe, I told smth wrong. Sorry, my english is not perfect ) I'll try again =))
The thing is not in building one plugin with another. The problem is sbt has a dependency on com.typsafe.config of version 1.2.0. And this library is loaded with ClassLoader of sbt itself. Play itself (not a plugin) uses more modern version and is incompatible with 1.2.0. I tried to create the plugin for sbt that uses play (not play plugin - and no matter what for )) ) but ran into an issue: it doesn't matter, what version of library I have in dependency of sbt plugin - if it is in sbt's dependencies or other plugin's dependencies with another version - it could be loaded first and break my one.
So, if you suppose it's OK that libraries of plugins may interfere and lead to conflicts - and there is no need for separate classloaders - that's your right ) But maybe you can update sbt's internal dependency to typesafe config's 1.3.1 version?
Thank you.

**Comment by dwijnand** [comment]
hey @kanischev,

I think what you're reporting is a duplicate of #2796.

I've created #4011 to track updating to typesafe config 1.3.1.

**Comment by kanischev** [comment]
hey @dwijnand 
Thank you. You're right. #2796 covers the same thing.

**Comment by tkubicz** [comment]
Hi @kanischev,

Yesterday, in our plugin we've encountered exactly the same problem with `typesafe config`. We've found a workaround for this issue using sbt-assembly shading feature ([source](https://manuzhang.github.io/2016/10/15/shading.html)) by renaming packages `"com.typesafe.**"` into some other packages.

I hope this help anyone who encounter this problem in the future until SBT gives us ability to use separate classloaders for plugins.
